### PR TITLE
Fix `local_fill_sink` rewrite for multiple output Elemwise Ops

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description

Reported by @tomicapretto in https://github.com/pymc-devs/pymc/pull/7315#pullrequestreview-2057678377

The rewrite implicitly assumed Elemwise nodes have a single output which is not true. The reported issue involved the gradient of BetaIncGrad which includes a ScalarLoop Op with multiple outputs.

The changes get rid of the eager sink at the local node rewriter level. This was actually not working because the nested replacements referenced variables that were never part of the original fgraph and those replacements were being ignored altogether. Instead we wrap the rewrite in an in2out that will safely achieve the intended behavior.

The new test works as a regression for the bug, in that if we were to call the old rewrite (or just try to compile the graph) it would lead to the reported issue when finding the MultiOutput Op.

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/tensor/rewriting/basic.py [hunk 1]
```diff
@@ -320,56 +321,52 @@ def dimshuffled_alloc(i):
...
     for inp in node.inputs:
         if inp.owner and inp.owner.op == fill:
-            models.append(inp.owner.inputs[0])
-            inputs.append(inp.owner.inputs[1])
+            a, b = inp.owner.inputs
+            if b.type.dtype != inp.dtype:
+                # The input was implicitly casted by the fill operation
+                b = b.cast(inp.dtype)
+            models.append(a)
+            inputs.append(b)
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/basic.py:342]: This was also a potential source of bugs in the old rewrite. Ops may behave fundamentally different if the input types change so we shouldn't let that happen

File: pytensor/tensor/rewriting/basic.py [hunk 2]
```diff
@@ -320,56 +321,52 @@ def dimshuffled_alloc(i):
...
 
-@register_canonicalize("shape_unsafe")
 @node_rewriter([Elemwise])
 def local_fill_sink(fgraph, node):
     """
     f(fill(a, b), fill(c, d), e) -> fill(c, fill(a, f(b, d, e)))
     f need to be an elemwise that isn't a fill.
     """
-    if not hasattr(node, "op") or not isinstance(node.op, Elemwise) or node.op == fill:
+    if isinstance(node.op.scalar_op, Second):
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/basic.py:330]: The extra checks were only needed for the recursive call of this rewrite. A default call will never call it on an node without Op that is not already Elemwise

**Comment by ricardoV94** [comment]
For reference fill is identical to `np.broadcast_arrays(x, y)[1]`, that is it broadcasts y to the broadcasted shape of x and y

**Review by tomicapretto — COMMENTED** [review]
I'm not able to say anything about the changes here because I'm not familiar with the parts being modified. But I can say this indeed fixes the example I shared on the linked PR, so this is good.

**Pull request merged**